src/modules/uORB/SubscriptionBlocking.hpp

```diff
@@ -0,0 +1,136 @@
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#pragma once
+
+#include "SubscriptionCallback.hpp"
+
+#include <containers/LockGuard.hpp>
+#include <px4_time.h>
+
+namespace uORB
+{
+
+// Subscription with callback that schedules a WorkItem
+template<typename T>
+class SubscriptionBlocking : public SubscriptionCallback
+{
+public:
+	/**
+	 * Constructor
+	 *
+	 * @param meta The uORB metadata (usually from the ORB_ID() macro) for the topic.
+	 * @param interval_ms The requested maximum update interval in milliseconds.
+	 * @param instance The instance for multi sub.
+	 */
+	SubscriptionBlocking(const orb_metadata *meta, uint8_t interval_ms = 0, uint8_t instance = 0) :
+		SubscriptionCallback(meta, interval_ms, instance)
+	{
+	}
+
+	virtual ~SubscriptionBlocking()
+	{
+		pthread_mutex_destroy(&_mutex);
+		pthread_cond_destroy(&_cv);
+	}
+
+	void call() override
+	{
+		LockGuard lg{_mutex};
+
+		// signal immediately if no interval, otherwise only if interval has elapsed
+		if ((_interval_us == 0) || (hrt_elapsed_time(&_last_update) >= _interval_us)) {
+			pthread_cond_signal(&_cv);
+		}
+	}
+
+	/**
+	 * Copy the struct if updated.
+	 * @param data The data reference where the struct will be copied.
+	 * @return true only if topic was updated and copied successfully.
+	 */
+	bool updateBlocking(T &data)
+	{
+		LockGuard lg{_mutex};
+
+		if (!_registered) {
+			registerCallback();
+		}
+
+		bool ret = false;
+
+		if (updated()) {
+			// copy immediately if updated
+			ret = copy(&data);
+
+		} else {
+			// otherwise wait with timeout based on interval
+			struct timespec ts {};
+			int ret_gettime = px4_clock_gettime(CLOCK_REALTIME, &ts);
```

Comment:
Why `CLOCK_REALTIME`?

Answering my own question: I guess the problem is that we don't have `pthread_condattr_setclock`. That's confusing.